### DISCUSSION TITLE
Narrowing IN with unions of TypedDict.

### DISCUSSION THREAD

**Discussion by JanEricNitschke** [microsoft/pyright#5070] [General]
I currently have a situation like below where i have multiple TypedDicts with (overlapping) keys of similar structure and i need to extract values from them. However i would like to avoid writing separate functions for each dict that do the same thing.

So i was trying to do something like this:

```
from typing import Literal, overload, TypedDict, LiteralString

class Test1(TypedDict)
    test1: str

class Test2(TypedDict):
    test2: str

@overload
def _get_actor_key(
    actor: Literal["1"],
    game_action: Test1,
) -> str:
    ...

@overload
def _get_actor_key(
    actor: Literal[""], game_action: Test2
) -> str:
    ...


def _get_actor_key(
    actor: Literal["1","2"],
    game_action: Test1 | Test2,
) -> str:
    actor_name = "test"+actor
    if actor_name in game_action:
        return game_action[actor_name]
    return ""
```

If i replace the type hints in the implementation with Any i get the desired behaviour of pyright rejecting invalid function calls like `_get_action_key("1", Test2)`.

But i can not managed to type hint the body with anything except any without it getting rejected. If i do the below it complains that "test1" is not a key for Test2 and vice versa. Is there any reasonable chance that something like this could be added to pyright or is there any way i could type hint this better or is just using Any my best bet here?

**Comment by erictraut** [comment]
Type narrowing doesn't apply in this case. The type of `game_action` is `Test1 | Test2`. Its type cannot be further narrowed based on the condition `actor_name in game_action`.

I would implement this as follows:

```python
def _get_actor_key(actor: Literal["1", "2"], game_action: Test1 | Test2) -> str:
    return game_action.get("test" + actor, "")
```
